File: kotlin/src/main/kotlin/problems1_100/LongestCommonPrefix.kt
package problems1_100

enum class LongestCommonPrefix {
    DEFAULT {
        override fun longestCommonPrefix(strs: Array<String>): String {
            if (strs.isEmpty()) {
                return ""
            }
            if (strs.size == 1) {
                return strs.first()
            }
            val longestStr = strs.maxBy { it.length }
            var end = -1
            loop@ for (i in longestStr.indices) {
                val c = longestStr[i]
                for (j in strs.indices) {
                    val str = strs[j]
                    if (i > str.lastIndex || c != str[i]) {
                        break@loop
                    } else if (j == strs.lastIndex) {
                        end = i
                    }
                }
            }
            return longestStr.substring(0, end + 1)
        }
    },
    ;

    abstract fun longestCommonPrefix(strs: Array<String>): String
}

File: javascript/src/problems1_100/longest-common-prefix.js
class Problem {
  longestCommonPrefix(strs) {
    throw new Error(`Not implemented (${strs}).`);
  }
}

class Default extends Problem {
  longestCommonPrefix(strs) {
    if (strs.length === 0) {
      return '';
    }
    if (strs.length === 1) {
      return strs[0];
    }
    const longestStr =
        strs.reduce((previous, current) => current.length > previous.length ? current : previous);
    let end = -1;
    loop:
        for (let i = 0; i < longestStr.length; i++) {
          const c = longestStr[i];
          for (let j = 0; j < strs.length; j++) {
            const str = strs[j];
            if (i > str.length - 1 || c !== str[i]) {
              break loop;
            } else if (j === strs.length - 1) {
              end = i;
            }
          }
        }
    return longestStr.slice(0, end + 1);
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;
